uportal-war/src/main/java/org/jasig/portal/portlets/dynamicskin/FileSystemDynamicSkinService.java

```diff
@@ -249,4 +250,33 @@ public String calculateTokenForCurrentSkin(PortletRequest request) {
         return Integer.toString(hash);
     }
 
+    @Override
+    public SortedSet<String> getSkinNames(PortletRequest request) {
+        // Context to access the filesystem
+        PortletContext ctx = request.getPortletSession().getPortletContext();
+
+        // Determine the full path to the skins directory
+        String skinsFilepath = ctx.getRealPath(rootFolder);
+
+        // Create File object to access the filesystem
+        File skinsDir = new File(skinsFilepath);
```

Comment:
I just looked into it.  The javascript (up-skin-selector.js) requests the XML file and parses it itself.  The back end does not do any parsing of the file.